# Put Option Simulation
---


# Question
---
* If you own a stock, buying a put option on the stock will greatly reduce your risk. This is the idea behind portfolio insurance. To illustrate, consider a stock that currently sells for $56 and has an annual volatility of 30%. Assume the risk-free rate is 8%, and you estimate that the stock’s annual growth rate is 12%.

* Suppose you own 100 shares of this stock. Use simulation to estimate the probability distribution of the percentage return earned on this stock during a one-year period.
  
* Now suppose you also buy a put option for $238 on the stock. The option has an exercise price of $50 and an exercise date one year from now. Use simulation to estimate the probability distribution of the percentage return on your portfolio over a one-year period. Can you see why this strategy is called a portfolio insurance strategy?

* Use simulation to show that the put option should, indeed, sell for about $238.

# Problem Breakdown
---
1. Simulate the stock price after one year to estimate the distribution of percentage return on the stock.
2. Incorporate the put option into the simulation to see how it affects the distribution of returns.
3. Estimate the fair price of the put option using simulation.

4. First, let's define some variables based on the given information:

- Initial stock price $S_0 = \$56$
- Annual volatility $\sigma = 30\%$
- Risk-free rate $r = 8\%$
- Stock’s annual growth rate $\mu = 12\%$
- Number of shares $N = 100$
- Put option exercise price $K = \$50$
- Option cost $C_{\text{option}} = \$238$

We'll simulate the stock price after one year using the geometric Brownian motion model, which is a common model for stock price dynamics. The formula for a future stock price after time $t$ is:

$$S_t = S_0 e^{(\mu - \frac{1}{2} \sigma^2)t + \sigma \sqrt{t} Z}$$

where $Z$ is a random variable from the standard normal distribution.


# Simulating Stock Price

In [4]:
import numpy as np

# Given variables
S0 = 56  # Initial stock price
mu = 0.12  # Drift (annual growth rate)
sigma = 0.30  # Volatility
r = 0.08  # Risk-free rate
N_shares = 100  # Number of shares
T = 1  # Time in years
n_simulations = 10000  # Number of simulations

# Simulate end-of-year stock prices
np.random.seed(0)  # For reproducibility
Z = np.random.randn(n_simulations)
ST = S0 * np.exp((mu - 0.5 * sigma**2) * T + sigma * np.sqrt(T) * Z)

# Calculate the percentage return on the stock
percentage_return_stock = ((ST - S0) / S0) * 100

# Results
print("Simulated percentage return on the stock")
print("Mean return: {:.2f}%".format(np.mean(percentage_return_stock)))
print("Standard deviation of return: {:.2f}%".format(np.std(percentage_return_stock)))


Simulated percentage return on the stock
Mean return: 12.01%
Standard deviation of return: 34.03%


# Incorporating the Put Option

In [5]:
# Calculate portfolio value with put option
K = 50  # Exercise price of the put option
C_option = 238  # Cost of the put option
option_payoff = np.maximum(K - ST, 0) * N_shares - C_option
portfolio_value_with_option = ST * N_shares + option_payoff - S0 * N_shares - C_option

# Calculate the percentage return on the portfolio
percentage_return_portfolio = (portfolio_value_with_option / (S0 * N_shares)) * 100

# Part 2 results
print("\nSimulated percentage return on the portfolio with put option")
print("Mean return: {:.2f}%".format(np.mean(percentage_return_portfolio)))
print("Standard deviation of return: {:.2f}%".format(np.std(percentage_return_portfolio)))



Simulated percentage return on the portfolio with put option
Mean return: 7.28%
Standard deviation of return: 29.88%


# Portfolio Insurance Strategy Findings
---
### Mean Return:

* Without the put option, the mean return on the stock is higher (12.01%) compared to the portfolio with the put option (7.28%). This is expected because the put option comes at a cost $238, which reduces the overall return on the portfolio.
---
### Standard Deviation of Return:

* The standard deviation of returns, which measures the risk or volatility of the investment, is lower for the portfolio with the put option (29.88%) compared to the stock alone (34.03%). This reduction in volatility is a key feature of insurance—by accepting a lower mean return, you're also reducing your exposure to downside risk.
The essence of a portfolio insurance strategy lies in its ability to limit losses without entirely forfeiting the potential for gains.
---
### Simulation:

* Without the put option, you're fully exposed to the stock's volatility, leading to a higher potential return but with greater risk.
With the put option, you've effectively set a floor (the exercise price of 50) on how much you can lose over the year. If the stock's value drops below this price, the put option will offset the losses beyond this point. This safety net comes at the cost of the option premium ($238), which is why the mean return on the insured portfolio is lower.
* By buying a put option, you're insuring the portfolio against significant losses, hence the term "portfolio insurance." This strategy is particularly appealing to investors who seek to participate in the potential upside of the stock market while protecting themselves against severe downturns.

# Estimating the Put Option Price
---
To estimate the fair price of the put option, we use the risk-neutral pricing formula. In a risk-neutral world, we replace the drift $\mu$ with the risk-free rate $r$ when simulating the stock prices. Then, we calculate the present value of the expected payoff of the put option.


In [6]:
# Simulate end-of-year stock prices under risk-neutral measure
ST_risk_neutral = S0 * np.exp((r - 0.5 * sigma**2) * T + sigma * np.sqrt(T) * Z)
option_payoff_risk_neutral = np.maximum(K - ST_risk_neutral, 0)
option_price_simulated = np.exp(-r * T) * np.mean(option_payoff_risk_neutral) * N_shares

# Part 3 results
print("\nEstimated price of the put option")
print("Simulated option price: ${:.2f}".format(option_price_simulated))



Estimated price of the put option
Simulated option price: $239.66


# Put Option Pricing
---
* Based on our simulation, the option price should have been priced at 239.66 and we were able to purchase it for $238. Meaning we got a slightly better deal on our downside protection in this case.